test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java

```diff
@@ -0,0 +1,214 @@
+                        executor.execute(task);
+                    }
+                    status = engine.getHandshakeStatus();
+                    break;
+                case FINISHED:
+                    break;
+                case NOT_HANDSHAKING:
+                    break;
+                default:
+                    throw new IllegalStateException("Invalid SSL status: " +
+                        status);
+            }
+        }
+        return status;
+    }
+
+    HandshakeStatus send(SSLEngine engine, ByteBuffer src, ByteBuffer dst)
+        throws SSLException {
+        SSLEngineResult status = engine.wrap(src, dst);
+        dst.flip();
+        return status.getHandshakeStatus();
+    }
+
+    HandshakeStatus receive(SSLEngine engine, ByteBuffer src, ByteBuffer dst)
+        throws SSLException {
+        SSLEngineResult status = engine.unwrap(src, dst);
+        dst.flip();
+        return status.getHandshakeStatus();
+    }
+
+    ReadOnlyEngine(SSLContext sslc, String ciphersuite) throws Exception {
+        System.err.println("==== Test Protocol: " + sslc.getProtocol() +
+            ", CipherSuite: " + ciphersuite);
+        KeyStore ks = KeyStore.getInstance("PKCS12");
+        KeyStore ts = KeyStore.getInstance("PKCS12");
+
+        ks.load(new FileInputStream(keyFilename), passwd);
+        ts.load(new FileInputStream(trustFilename), passwd);
+
+        KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
+        kmf.init(ks, passwd);
+        TrustManagerFactory tmf = TrustManagerFactory.getInstance("SunX509");
+        tmf.init(ts);
+
+        sslc.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
+        client = sslc.createSSLEngine("client", 1);
+        client.setUseClientMode(true);
+        server = sslc.createSSLEngine("server", 2);
+        if (ciphersuite != null) {
+            server.setEnabledCipherSuites(new String[] { ciphersuite });
+        }
+        server.setUseClientMode(false);
+
+        SSLSession session = server.getSession();
+        int maxData = session.getPacketBufferSize();
+
+        ByteBuffer in = ByteBuffer.allocate(maxData);
+        ByteBuffer out = ByteBuffer.allocate(maxData);
+
+        HandshakeStatus statusClient, statusServer;
+        client.beginHandshake();
+        server.beginHandshake();
+
+        // Do TLS handshake
+        do {
+            statusClient = doHandshake(client, out, in);
+            statusServer = doHandshake(server, in, out);
+        } while (statusClient != HandshakeStatus.NOT_HANDSHAKING ||
+            statusServer != HandshakeStatus.NOT_HANDSHAKING);
+
+        // Read NST
+        in.clear();
+        receive(client, out, in);
+
+        System.out.println("done");
+
+        // Send bytes from the client and make sure the server receives the same
+        in.clear();
+        out.clear();
+        String testString = "ASDF";
+        in.put(testString.getBytes()).flip();
```

Comment:
If you're going to convert back from UTF_8 later, you should probably convert using getBytes(UTF_8) here.

Reply:
setting the input to UTF8 isn't a concern.  The latter line to decode it changes it from using the ByteBuffer.toString() to the contents of the ByteBuffer in a String.

Reply:
You could use the default charsets for encoding and decoding.  i.e. 

        in.clear();
        receive(server, out.asReadOnlyBuffer(), in);
        byte[] ba = new byte[in.remaining()];
        in.get(ba);
        String testResult = new String(ba);